pkg/controllers/kdd-ipam/controller.go

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kddipam
+
+import (
+	"context"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	uruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/projectcalico/kube-controllers/pkg/config"
+	"github.com/projectcalico/kube-controllers/pkg/controllers/controller"
+	bapi "github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+)
+
+// Controller implements the Controller interface.  It is responsible for monitoring
+// kubernetes nodes and responding to delete events by removing them from the Calico datastore.
+type Controller struct {
+	ctx       context.Context
+	informer  cache.Controller
+	indexer   cache.Indexer
+	client    client.Interface
+	clientset *kubernetes.Clientset
+	bc        bapi.Client
+}
+
+type allocation struct {
+	handle     string
+	attributes map[string]string
+}
+
+// NewController Constructor for Controller
+func NewController(ctx context.Context, clientset *kubernetes.Clientset, c client.Interface, cfg *config.Config) controller.Controller {
+	// Used to extract the backend client we need for IPAM calls.
+	type accessor interface {
+		Backend() bapi.Client
+	}
+
+	nc := &Controller{
+		ctx:       ctx,
+		client:    c,
+		bc:        c.(accessor).Backend(),
+		clientset: clientset,
+	}
+
+	return nc
+}
+
+// Run starts the node controller. It does start-of-day preparation
+// and then launches worker threads. We ignore reconcilerPeriod and threadiness
+// as this controller does not use a cache and runs only one worker thread.
+func (c *Controller) Run(threadiness int, reconcilerPeriod string, stopCh chan struct{}) {
+	defer uruntime.HandleCrash()
+
+	log.Info("Starting Kubernetes IPAM controller")
+
+	log.Info("Kubernetes IPAM controller is now running")
+	go c.run()
+
+	<-stopCh
+	log.Info("Stopping Kubernetes IPAM controller")
+}
+
+func (c *Controller) run() {
+	for {
```

Comment:
Should probably switch to a kick-based model.

At the least, we only need to do this every 10 minutes or so. Better yet would be to only perform a sync loop whenever a node is deleted, with a fallback of like an hour so.

Reply:
Yeah, I'm a little confused. The Indexer/Informer variables are defined on the Controller struct but I don't seem them actually being used.